consensus/src/consensus_observer/payload_store.rs

```diff
@@ -4,277 +4,877 @@
-                mem::swap(entry.get_mut(), &mut status);
-
-                // If the status was originally requested, send the payload to the listener
-                if let BlockPayloadStatus::Requested(payload_sender) = status {
-                    if payload_sender.send(block_transaction_payload).is_err() {
-                        error!(LogSchema::new(LogEntry::ConsensusObserver)
-                            .message("Failed to send block payload to listener!",));
-                    }
-                }
-            },
-            Entry::Vacant(entry) => {
-                // Insert the block payload directly into the payload store
-                entry.insert(BlockPayloadStatus::Available(block_transaction_payload));
-            },
+        // Verify that the number of payloads doesn't exceed the maximum
+        let max_num_pending_blocks = self.consensus_observer_config.max_num_pending_blocks as usize;
+        if self.block_payloads.lock().len() >= max_num_pending_blocks {
+            warn!(
+                LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                    "Exceeded the maximum number of payloads: {:?}. Dropping block: {:?}!",
+                    max_num_pending_blocks, block_payload.block,
+                ))
+            );
+            return; // Drop the block if we've exceeded the maximum
         }
+
+        // Create the new payload status
+        let epoch_and_round = (block_payload.block.epoch(), block_payload.block.round());
+        let payload_status = if verified_payload_signatures {
+            BlockPayloadStatus::AvailableAndVerified(block_payload)
+        } else {
+            BlockPayloadStatus::AvailableAndUnverified(block_payload)
+        };
+
+        // Insert the new payload status
+        self.block_payloads
+            .lock()
+            .insert(epoch_and_round, payload_status);
     }
 
-    /// Removes the given pipelined blocks from the payload store
-    pub fn remove_blocks(&self, blocks: &[Arc<PipelinedBlock>]) {
-        let mut block_transaction_payloads = self.block_transaction_payloads.lock();
-        for block in blocks.iter() {
-            block_transaction_payloads.remove(&block.id());
-        }
+    /// Removes all blocks up to the specified epoch and round (inclusive)
+    pub fn remove_blocks_for_epoch_round(&self, epoch: u64, round: Round) {
+        // Determine the round to split off
+        let split_off_round = round.saturating_add(1);
+
+        // Remove the blocks from the payload store
+        let mut block_payloads = self.block_payloads.lock();
+        *block_payloads = block_payloads.split_off(&(epoch, split_off_round));
+    }
+
+    /// Removes the committed blocks from the payload store
+    pub fn remove_committed_blocks(&self, committed_blocks: &[Arc<PipelinedBlock>]) {
+        // Identify the highest epoch and round for the committed blocks
+        let (highest_epoch, highest_round) = committed_blocks
```

Comment:
it should be the last one without needing to iterate

Reply:
Aah, good spot -- yeah, the blocks should already be verified here. 😄